contracts/utils/SelfMulticall.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./interfaces/ISelfMulticall.sol";
+
+/// @notice Contract that enables calls to the inheriting contract to be
+/// batched
+/// @dev Refer to OpenZeppelin's Multicall.sol for a similar implementation
+contract SelfMulticall is ISelfMulticall {
+    /// @notice Batches calls to the inheriting contract and reverts if at
+    /// least one of the batched calls reverts
+    /// @dev Bubbles up the raw returndata if a call reverts, leaving decoding
+    /// to the user
+    /// @param data Array of calldata of batched calls
+    /// @param returndata Array of returndata of batched calls
+    function multicall(bytes[] calldata data)
+        external
+        override
+        returns (bytes[] memory returndata)
+    {
+        returndata = new bytes[](data.length);
+        for (uint256 i = 0; i < data.length; i++) {
+            bool success;
+            (success, returndata[i]) = address(this).delegatecall(data[i]); // solhint-disable-line avoid-low-level-calls
+            // Adapted from OpenZeppelin's Address.sol
+            if (!success) {
+                if (returndata.length > 0) {
+                    // solhint-disable-next-line no-inline-assembly
+                    assembly {
+                        let returndata_size := mload(returndata)
+                        revert(add(32, returndata), returndata_size)
```

Comment:
Do you feel it would be to risky trying to decode the revert string from the `returndata`? Like in here https://github.com/boringcrypto/BoringSolidity/blob/master/contracts/BoringBatchable.sol#L28

Reply:
So the Boring implementation assumes `returndata` is a `string` that is ABI-encoded into `bytes`. The OpenZeppelin implementation assumes `returndata` is anything that is ABI-encoded into `bytes`. An additional improvement over that would be not making any assumptions about `returndata`. For example, `returndata` can be `0x1234` (as a result of a future fork), which is not a valid ABI-encoding because the bytelength is not a multiple of 32. The problem is there are a lot of other ways for a `bytes` to not be a valid ABI encoding (and not only because of length), and the only way to verify that is knowing the schema beforehand, which we don't with custom errors (with only string reverts we could have verified it by re-encoding it as in [here](https://github.com/api3dao/airnode-protocol-v1/blob/c7eae72bcd82292a05f7a72d9a81dc7f350974d0/contracts/dapis/DapiServer.sol#L507)).

So the best that can be done is making a best effort as in OZ. The thing to consider is what happens to the OZ implementation if `returndata` stops conforming to ABI (for example if it's `0x1234`). It then misinterprets the first 32 bytes of the `returndata` as the length of the data to be reverted with. In the case of `0x1234`, this will appear as `0x1234..............` (spilling over to the memory space to read the rest), which is a huge number. Then, this is what the call will revert with, and you will pay a lot of gas if you try to emit this as in RRP.

Then to use the OZ implementation, we would have to assume that `returndata` will always conform to contract ABI. The thing is, contract ABI is designed to represent calldata and returndata, so this is somewhat of a safe bet. I think searcher ExternalMulticall contracts should definitely use this, but I'm still not 100% convinced about adding it to AirnodeProtocol and DapiServer